AMBARI-25151 SystemD support for ambari-agent on SLES12 (benyoka)

benyoka opened the pull request:
## What changes were proposed in this pull request?

Added support for SystemD startup / shoutdown + workaround for DefaultTasksMax=512 limitation for SLES 12.x

## How was this patch tested?
- built **ambari-agent** with the -**Psuse12** flag
- compared the **suse12** rpm with the normal rpm build. The only difference is the expected `/usr/lib/systemd/system/ambari-agent.service` addition.
- Tested the rpm with SLES12.2 for the following scenarios:
   - ambari-agent running as **root**, HDFS running as **hdfs**
   - ambari-agent running as **root**, HDFS running as **root**
   - ambari-agent running as **ambari**, HDFS running as **hdfs**
   - ambari-agent running as **ambari**, HDFS running as **root**

Commit:
AMBARI-25151 SystemD support for ambari-agent on SLES12 (benyoka)

Commit:
AMBARI-25151 fix service unit file (benyoka)

Commit:
AMBARI-25151 fix service unit file (benyoka)

Commit:
AMBARI-25151 review comment (benyoka)

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ambari-Github-PullRequest-Builder/4921/
Test PASSed.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ambari-Github-PullRequest-Builder/4923/
Test FAILed.
Test FAILured.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ambari-Github-PullRequest-Builder/4922/
Test PASSed.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ambari-Github-PullRequest-Builder/4924/
Test PASSed.

kasakrisz reviewed (approved):
LGTM

hapylestat commented on ambari-agent/conf/unix/ambari-agent.service:
```diff
@@ -0,0 +1,40 @@
+
+[Unit]
+Description=Ambari Agent Service for SLES
+Before=runlevel2.target
```
runlevel 2 - is a system start without network
runlevel 4 - is not defined 

Why we require agent start on the stages when agent will not actually work anyways?

benyoka replied:
Thanks for spotting it. Basically I took the Centos generated service file as start, hence it is there. I will remove.

benyoka commented:
@g-boros please have look.

Commit:
AMBARI-25151 remove runlevel 2 and 4 (benyoka)

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ambari-Github-PullRequest-Builder/4926/
Test PASSed.

benyoka commented:
cc @lacikaaa

lacikaaa commented on ambari-agent/conf/unix/ambari-agent.service:
```diff
@@ -0,0 +1,38 @@
+
+[Service]
+Type=forking
+Restart=no
```
are you sure about this? currently CB deploys it with always

benyoka replied:
I would like to keep consistency with other systems and make the minimal change needed.

lacikaaa reviewed (commented):
`PIDFile=` is missing although systemd documentation suggests when using `Type=forking` it should be set

Commit:
AMBARI-25151 add pidfile, disable RemainAfterExit (benyoka)

benyoka commented:
Added **PIDFile**. Also changed **RemainAfterExit** to **no** so that systemd recognizes that the service is stopped (e.g. by killing it or by calling _ambari-agent stop_)

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ambari-Github-PullRequest-Builder/4930/
Test FAILed.
Test FAILured.

benyoka commented:
retest this please

lacikaaa reviewed (approved):
LGTM

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ambari-Github-PullRequest-Builder/4931/
Test PASSed.

benyoka commented:
Hey @hapylestat, pls let me know if you're ok with the PR.

yj7778826 commented:
why /usr/lib/systemd/system/ambari-agent.service will be installed on debian9? It will cause ambari-agent not autostart after system reboot.
